### PR TITLE
initial download-folder-selector interface

### PR DESCRIPTION
**Pull request by mcmonkey4eva**
Depends on https://github.com/comfyanonymous/ComfyUI/pull/4981 and https://github.com/comfyanonymous/ComfyUI/pull/4980

This is a followup from earlier PR https://github.com/Comfy-Org/ComfyUI_frontend/pull/569

-----

Adds a model download folder selector

If I use vanilla HTML Select it's naturally quite ugly
image

PrimeVue looks nicer:
image

~~But it's weirdly slow about updating when you select things and idk what's up with that. I think it's something to do with `v-model` and the indirection towards the actual folder path, but I'm not entirely sure.~~ fixed

It's also chonky and ugly regardless so I added a visibility checkbox
image
(Most users won't want to change from default in theory)

## TODO

- user setting for default folder path, or just generally persist selections
- ~~fix the select interface being all wonky~~ fixed
- ~~make the tests pass lol~~ the merge fixed it? lol

### CODE REVIEW

**Review by huchenlei — CHANGES_REQUESTED** [review]
Can you add at least a playwright test to verify the UI is shown correctly? You can add code to https://github.com/Comfy-Org/ComfyUI_devtools to perform test setup.

**Comment by mcmonkey4eva** [comment]
alright, now it has playwright testing on it, screenshots all look correctly. (Note the folder selector shows 'models' and 'output' in the screenshot, visually similar but not wrong)

**Comment by huchenlei** [comment]
> alright, now it has playwright testing on it, screenshots all look correctly. (Note the folder selector shows 'models' and 'output' in the screenshot, visually similar but not wrong)

The general rule of thumb for screenshot test is that we only do that do check litegraph canvas. For things that can be checked via inspecting DOM / evaluating JS, we do that instead to make test less fragile to subtle style changes and running locally.

**Review by huchenlei — CHANGES_REQUESTED** [review]
Please rewrite the test to avoid using screenshot expectations.

**Comment by mcmonkey4eva** [comment]
Kinda hurts not being able to do the thorough validation of screenshots but I get it, too much changeweight any time minor stylistic stuff is altered.

I rewrote the tests to exclude screenshots and instead check visibility and classnames to ensure they are reacting appropriately.

There's probably room to be more thorough than that? But that covers the key predictable functional parts, short of writing a full unit test that mocks the folder list and download apis to test end-to-end.

File: src/components/dialog/content/MissingModelsWarning.vue [hunk 1]
```diff
@@ -60,19 +82,22 @@
...
 
-const settingStore = useSettingStore()
-const allowedSources = settingStore.get(
-  'Comfy.Workflow.ModelDownload.AllowedSources'
-)
-const allowedSuffixes = settingStore.get(
-  'Comfy.Workflow.ModelDownload.AllowedSuffixes'
-)
+const showFolderSelect = ref(false)
+
```
Comment (huchenlei) [inline src/components/dialog/content/MissingModelsWarning.vue:94]: Why this becomes hard coded again? We do want to expose this configurability to cloud deployment platforms to download from custom endpoints.

**Review by huchenlei — CHANGES_REQUESTED** [review]
Please revert the hard coded sources and suffixes.

**Comment by mcmonkey4eva** [comment]
They need to _not_ be in user settings, and they should ideally not be something that gets corrupted by an accidental run of the playwright test.

If we want to expose server configurability, it should not be through user settings.

As noted earlier, the most ideal place imo would be an internal API on the Comfy core side that returns its actual accepted value list, and then the UI simply reads and works with that. (The UI is simply trying to preview what the server will determine, and atm serverside it's hardcoded)
If you agree with that, I can go PR that on the core side.

**Comment by huchenlei** [comment]
> They need to _not_ be in user settings, and they should ideally not be something that gets corrupted by an accidental run of the playwright test.
> 
> If we want to expose server configurability, it should not be through user settings.
> 
> As noted earlier, the most ideal place imo would be an internal API on the Comfy core side that returns its actual accepted value list, and then the UI simply reads and works with that. (The UI is simply trying to preview what the server will determine, and atm serverside it's hardcoded) If you agree with that, I can go PR that on the core side.

I think the playwright test side issue can be fixed by having extra logic backup server side user files and restore them after running all tests. Just for testability, they should not be hard-coded.

**Comment by mcmonkey4eva** [comment]
`Just for testability, they should not be hard-coded.` the tests work cleanly currently

**Comment by huchenlei** [comment]
> `Just for testability, they should not be hard-coded.` the tests work cleanly currently

Do you really want to expose localhost to default model download whitelist origin?

**Comment by mcmonkey4eva** [comment]
`Do you really want to expose localhost to default model download whitelist origin?`
Yes imo this is fine to allow `localhost` to render as if valid (though the server will not actually download from localhost)
(And note again we'll want to replace the list with server-read source either now or in the future)
(if you want it to be strictly correct before merging, I'm happy to make the core PR to provide the relevant internal API to do the list properly)

File: src/components/dialog/content/MissingModelsWarning.vue [hunk 2]
```diff
@@ -60,19 +82,22 @@
...
+const allowedSources = [
+  'https://civitai.com/',
+  'https://huggingface.co/',
+  'http://localhost:'
```
Comment (huchenlei) [inline src/components/dialog/content/MissingModelsWarning.vue]: At least document that localhost origin is for testing only.

Comment (mcmonkey4eva) [reply src/components/dialog/content/MissingModelsWarning.vue]: done

**Pull request merged**